app/models/concerns/host_info_providers/puppet_info.rb

```diff
@@ -0,0 +1,116 @@
+module HostInfoProviders
+  module PuppetInfo
+    extend ActiveSupport::Concern
+
+    included do
+      register_info_provider :puppet_info
+    end
+
+    def puppet_info
+      # Static parameters
+      param = parameters_info_hash
+
+      # Parse ERB values contained in the parameters
+      param = SafeRender.new(:variables => { :host => self }).parse(param)
+
+      info_hash = {}
+      info_hash['classes'] = classes_info_hash
+      info_hash['parameters'] = param
+      info_hash['environment'] = param["foreman_env"] if Setting["enc_environment"] && param["foreman_env"]
+
+      info_hash
+    end
+
+    def parameters_info_hash
+      params = {}
+      # maybe these should be moved to the common parameters, leaving them in for now
+      params["puppetmaster"] = puppetmaster
+      params["puppet_ca"]    = puppet_ca_server if SETTINGS[:unattended] && puppetca_exists?
+
+      params["foreman_env"]  = environment.to_s unless environment.nil? || environment.name.nil?
+
+      if Setting[:ignore_puppet_facts_for_provisioning]
+        params["ip"]  = ip
+        params["ip6"] = ip6
+        params["mac"] = mac
+      end
+
+      params.update(smart_variables)
+    end
+
+    def classes_info_hash
+      return [] if self.environment.nil?
+
+      return puppetclass_parameters if Setting[:Parametrized_Classes_in_ENC] && Setting[:Enable_Smart_Variables_in_ENC]
+
+      puppetclasses_names
+    end
+
+    def puppetclass_parameters
+      classifier = Classification::Base.new self
+      keys = PuppetclassLookupKey.includes(:environment_classes).parameters_for_class(puppetclass_ids, environment_id)
+      key_hash = hashed_class_keys(keys)
+      values = classifier.values_hash(keys)
+
+      klasses = {}
+      classes.each do |klass|
+        klasses[klass.name] = smart_class_params_for klass, key_hash, values, classifier
```

Comment:
nitpick, try to not use parenthesis for well-known APIs (Ruby, Rails & little else)  and use them otherwise, I don't know if you follow another convention or it's random but it gets a bit difficult to read